native/src/fairseq2n/data/sample_data_source.cc

```diff
@@ -170,9 +175,16 @@ sample_data_source::random_pipeline_index()
             rptr = mptr;
     }
 
+    // The binary search returns the index where cumsum[idx] >= sample and cumsum[idx-1] < sample
+    // If this pipeline is blocked, it means cumsum[idx] == cumsum[idx-1], which is a
+    // degenerate interval. We need to search forward to find the pipeline that actually
+    // owns this probability mass (the next unblocked one with cumsum[idx] > cumsum[idx-1])
+    while (lptr < weight_cumsums_.size() && is_blocked_[lptr]) {
```

Comment:
As discussed offline, this can be reduced to:

```python

if (!allow_repeats_) {
  while (is_epoch_done_[lptr])
    lptr++;
}
```

I would also put `lptr < weight_cumsums_.size()` optionally as an assert statement since due to `are_all_done()` check, `lptr` should never move beyond the range of `is_epoch_done_`.